pennylane/capture/transforms.md

```diff
@@ -0,0 +1,370 @@
+        of the current level"""
+        assert False  # must override
+
+    def process_primitive(self, primitive, tracers, params):
+        """Evaluate a primitive based on what the transformation is supposed
+        to do. Here, we can bind new primitives, which might be interpreted by
+        Traces lower in the stack."""
+        assert False  # must override
+```
+
+`pure` and `lift` are for boxing up values in `Tracer`s, and `process_primitive` is the callback we
+use to transform primitives. Multiple instances of `Trace` might be created and discarded during
+application of a transformation, whereas only a single `MainTrace` instance is created per
+application of a transformation.
+
+```python
+class Tracer:
+    _trace: Trace
+
+    __array_priority__ = 1000
+
+    @property
+    def aval(self):
+        """Abstract value of the Tracer"""
+        assert False  # must override
+
+    def full_lower(self):
+        return self  # default implementation
+
+    def __getattr__(self, name):
+        try:
+            return getattr(self.aval, name)
+        except AttributeError:
+            raise AttributeError(f"{self.__class__.__name__} has no attribute {name}")
+```
+
+A `Tracer` carries an abstract value, and the rest is up to the transformation.
+
+Now, let's set up minimal versions of other JAX helper functions that are used during
+interpretation, namely `jax.core.bind_with_trace` (which is called by `Primitive.bind`),
+`jax.core.find_top_trace` (to figure out which `MainTrace`, or interpreter, should handle
+the current primitive), `jax.core.full_raise` (to ensure that values are boxed up in the
+top `Trace`'s `Tracer`s), and `jax.core.full_lower` (an optional optimization to unbox
+values out of `Tracer`s as much as possible).
+
+```python
+from typing import Sequence, Union
+
+def find_top_trace(xs: Sequence[Union[Any, Tracer]]) -> Trace:
+    """Find the MainTrace corresponding to the highest level interpreter that should be used for
+    the given inputs, and return a Trace corresponding to the trace_type of that MainTrace"""
+    top_main = max(
+        (x._trace.main for x in xs if isinstance(x, Tracer)),
+        default=trace_stack[0],
+        key=lambda main: main.level,
+    )
+
+    return top_main.trace_type(top_main)
+```
+`find_top_trace` returns the highest-level interpreter associated with the `Tracer`s on its inputs,
+and otherwise returns the interpreter at the bottom of the stack. This last interpreter does the
+standard evaluation on the primitive we are binding (more on that later). Note that we are only
+applying an interpreter when the input arguments to a primitive are boxed in a `Tracer`
+corresponding to that interpreter. This is an optimization that JAX uses to skip usage of
+unnecessary interpreters, so that transformations follow data-dependence (except for the special
+level-0 interpreter, which interprets everything). This data-dependence will cause issues with
```

Comment:
```suggestion
unnecessary interpreters, so that transformations follow data dependence (except for the special
level-0 interpreter, which interprets everything). This data dependence will cause issues with
```